### PR TITLE
Workspace start time metrics

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
The goal of this pr is to provide metrics that are describing how fast workspace starting or failing to start. They are divided into two parts. More informative [2][3] represented as a heat map and less informative [1] represented as a graph. The idea is that on a graph [1] , that is open by default, admin can see general trends, and on the graph [2][3] he can receive more details.

## Details
[1] - represent 5min average time of workspace successful start or fail.
[2][3] -   represent 1h average time of workspace successful start or fail dived by on rows. Where row represents some period of time. For example: 22-28 - means the number of workspaces started/failed in the period from 22 to 28 seconds

Знімок екрана  о 15 42 36
Знімок екрана  о 15 42 53

### What issues does this PR fix or reference?
n/a

#### Release Notes
n/a

#### Docs PR
n/a

### CODE REVIEW

**Review by metlos — APPROVED** [review]

**Comment by skabashnyuk** [comment]
ci-test

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            .sla(Duration.ofSeconds(10))
+            .minimumExpectedValue(Duration.ofSeconds(10))
+            .maximumExpectedValue(Duration.ofMinutes(15))
+            .register(registry);
+
+    // only subscribe to the event once we have the counters ready
+    eventService.subscribe(
+        event -> {
+          if (event.getPrevStatus() == WorkspaceStatus.STOPPED
+              && event.getStatus() == WorkspaceStatus.STARTING) {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: When is it possible that PrevStatus is not `STOPPED` but status is `STARTING`?

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            workspaceStartTime.put(event.getWorkspaceId(), System.currentTimeMillis());
+          } else if (event.getPrevStatus() == WorkspaceStatus.STARTING) {
+            Long startTime = workspaceStartTime.remove(event.getWorkspaceId());
+            if (startTime != null) {
+              if (event.getStatus() == WorkspaceStatus.RUNNING) {
+                successTimer.record(Duration.ofMillis(System.currentTimeMillis() - startTime));
+              } else if (event.getStatus() == WorkspaceStatus.STOPPED) {
```
Comment (sleshchenko) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: Note that you will register failing even if a user interrupts workspace start.

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            .description("The time of workspace start")
+            .tags(withStandardTags("result", "fail"))
+            .publishPercentileHistogram()
+            .sla(Duration.ofSeconds(10))
+            .minimumExpectedValue(Duration.ofSeconds(10))
+            .maximumExpectedValue(Duration.ofMinutes(15))
+            .register(registry);
+
+    // only subscribe to the event once we have the counters ready
+    eventService.subscribe(
```
Comment (sleshchenko) [inline wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: Please consider a the following way to rewrite:
```java
    // only subscribe to the event once we have the counters ready
    eventService.subscribe(this::handleWorkspaceStatusChange, WorkspaceStatusEvent.class);
  }

  private void handleWorkspaceStatusChange(WorkspaceStatusEvent event) {
    if (event.getStatus() == WorkspaceStatus.STARTING) {
      workspaceStartTime.put(event.getWorkspaceId(), System.currentTimeMillis());
    } else if (event.getPrevStatus() == WorkspaceStatus.STARTING) {
      Long startTime = workspaceStartTime.remove(event.getWorkspaceId());
      if (startTime == null) {
        // this situation is possible when starting workspace is recovered or some bug happened
        LOG.warn("No workspace start time recorded for workspace {}", event.getWorkspaceId());
        return;
      }

      if (event.getStatus() == WorkspaceStatus.RUNNING) {
        successTimer.record(Duration.ofMillis(System.currentTimeMillis() - startTime));
      } else if (event.getStatus() == WorkspaceStatus.STOPPED) {
        failTimer.record(Duration.ofMillis(System.currentTimeMillis() - startTime));
      }
    }
  }
```

**Review by sleshchenko — APPROVED** [review]

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java (hunk 1 above)
Comment (skabashnyuk) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: in tests. Basically, it's "Fool protection"

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java (hunk 2 above)
Comment (skabashnyuk) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: I consider this an acceptable scenario.

File: wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java (hunk 3 above)
Comment (skabashnyuk) [reply wsmaster/che-core-api-metrics/src/main/java/org/eclipse/che/api/metrics/WorkspaceStartTrackerMeterBinder.java]: ok

**Pull request merged**